### PR TITLE
GH-39489: [C++][Parquet] ConvertedType TIME and TIMESTAMP should imply adjustedToUtc

### LINKED ISSUE #39489: [C++][Parquet] Timestamp conversion from Parquet to Arrow does not follow compatibility guidelines for convertedType
### Describe the bug, including details regarding any error messages, version, and platform.

The Parquet format [document](https://github.com/apache/parquet-format/blob/eb4b31c1d64a01088d02a2f9aefc6c17c54cc6fc/LogicalTypes.md?plain=1#L480-L485) indicates that converted types should have `isAdjustedToUtc=true` for backward compatibility, but the current C++ [implementation](https://github.com/apache/arrow/blob/37a8bf04bc713858a5b247d4424c1e8505e61947/cpp/src/parquet/arrow/schema_internal.cc#L92-L93) sets this to `false` for converted types.

The [Rust implementation](https://github.com/apache/arrow-rs/blob/a61e824abdd7b38ea214828480430ff2a13f2ead/parquet/src/arrow/schema/primitive.rs#L211-L239) already follows the spec as described and the Go implementation has #39467 open to correct this.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

TLDR: When reading from a Parquet file with legacy timestamp not written by arrow, `isAdjustedToUTC` would be ignored during read. And when filtering a file like this, filtering would not work.

When casting from a "deprecated" parquet 1.0 ConvertedType, a timestamp should be force adjustedToUtc.

For the parquet standard part. Parquet has a ConvertedType for legacy timestamp, the legacy timestamp **do not** having a `adjustedToUtc` flag. So, for forward compatible, when reading it we need to regard it as `adjustedToUtc` ( A UTC Timestamp). See [1] [2].

In our implementation, we use a `(LogicalType, PhysicalType)` pair, `LogicalType` it's a wrapper for parquet thrift `ConvertedType` and `LogicalType`, the code is listed in [3] : 
1. During write, we always use `FieldToNode` and `TimestampLogicalTypeFromArrowTimestamp` [4] to generate a `TimestampLogicalType`, and generate `isAdjustedUtc` by `::arrow::TimestampType::timezone()`.  Also, the arrow type will be a `ARROW:schema` extended key-value, and written in parquet file. During write, we'll both set logicaltype and convertType. `arrow_properties.coerce_timestamps_enabled()` and `arrow_properties.support_deprecated_int96_timestamps()` control the behavior.
2. During read, we'll first parse parquet thrift and generate a `LogicalType`[5]. And then, `FromParquetSchema`[6] would be called for arrow type.

in 1-5, code works well, because LogicalType parsing works well. However, [6] might causing some problem:
1. If `ARROW:schema` extended is provided, nothing is wrong
2. **Otherwise, timestamp would be wrong because it ignore isAdjustedUtc!**. So when it read legacy Parquet file not written by arrow, it would not have right type.

Further more, this could happening when dataset filtering happens.

[1] https://github.com/apache/parquet-format/blob/eb4b31c1d64a01088d02a2f9aefc6c17c54cc6fc/LogicalTypes.md?plain=1#L480-L485
[2] https://github.com/apache/parquet-format/blob/eb4b31c1d64a01088d02a2f9aefc6c17c54cc6fc/LogicalTypes.md?plain=1#L308
[3] https://github.com/apache/arrow/blob/main/cpp/src/parquet/schema.cc#L128-L283
[4] https://github.com/apache/arrow/blob/main/cpp/src/parquet/arrow/schema.cc#L140
[5] https://github.com/apache/arrow/blob/main/cpp/src/parquet/schema.cc#L443
[6] https://github.com/apache/arrow/blob/main/cpp/src/parquet/arrow/schema.cc#L1038

### What changes are included in this PR?

ConvertedType TIME and TIMESTAMP should imply adjustedToUtc when casting to arrow

### Are these changes tested?

Already

### Are there any user-facing changes?

User might got some behavior change during read the legacy timestamp. But most of time, arrow would casting it.

* Closes: #39489

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
> Are these changes tested? Already

Given that you change something, but not any tests, that seems to indicate that this isn't actually tested?

> User might got some behavior change during read the legacy timestamp. But most of time, arrow would casting it.

Can we try to be more explicit about the exact impact of this change? In which cases does it change behaviour for a user? (I suppose it will change in some cases, otherwise why are we changing it?)

**Comment by mapleFU** [comment]
@jorisvandenbossche Sorry for being so short. I forgot to upload testing previously. I upload the description in this patch. It only happens when:
1. reading a file with legacy type not written by arrow
2. Filtering

File: cpp/src/parquet/arrow/schema_internal.cc [hunk 1]
```diff
@@ -90,7 +90,7 @@ Result<std::shared_ptr<ArrowType>> MakeArrowTime64(const LogicalType& logical_ty
 Result<std::shared_ptr<ArrowType>> MakeArrowTimestamp(const LogicalType& logical_type) {
   const auto& timestamp = checked_cast<const TimestampLogicalType&>(logical_type);
   const bool utc_normalized =
-      timestamp.is_from_converted_type() ? false : timestamp.is_adjusted_to_utc();
+      timestamp.is_from_converted_type() ? true : timestamp.is_adjusted_to_utc();
```
Comment (wgtmac) [inline cpp/src/parquet/arrow/schema_internal.cc:93]: CMIW, the change here is to align with cpp/src/parquet/types.cc where is_adjusted_to_utc of logical type is set true.

Comment (mapleFU) [reply cpp/src/parquet/arrow/schema_internal.cc:93]: Here, every caller except here is set to true.

By the way, I didn't find Int96 in the doc, would Int96 should having `is_adjusted_to_utc = true` ? (Rust impl set it false)

**Comment by mapleFU** [comment]
@jorisvandenbossche @pitrou @wgtmac Would you mind take a look?

File: cpp/src/parquet/arrow/schema_internal.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema_internal.cc:93]: Why not just
```cpp
  const bool utc_normalized = timestamp.is_adjusted_to_utc();
```

File: cpp/src/parquet/types.cc [hunk 2]
```diff
@@ -345,15 +345,19 @@ std::shared_ptr<const LogicalType> LogicalType::FromConvertedType(
     case ConvertedType::DATE:
       return DateLogicalType::Make();
     case ConvertedType::TIME_MILLIS:
```
Comment (pitrou) [inline cpp/src/parquet/types.cc:347]: Can you add a reference to https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype ?

Comment (mapleFU) [reply cpp/src/parquet/types.cc:347]: done

File: cpp/src/parquet/arrow/schema_internal.cc (hunk 1 above)
Comment (joellubi) [reply cpp/src/parquet/arrow/schema_internal.cc]: I think the following would be equivalent to the updated ternary expression:
```cpp
const bool utc_normalized = timestamp.is_from_converted_type() || timestamp.is_adjusted_to_utc();
```

Comment (mapleFU) [reply cpp/src/parquet/arrow/schema_internal.cc]: I've changed to:

```
  const bool utc_normalized = timestamp.is_adjusted_to_utc();
```

I think this is enough

Comment (wgtmac) [reply cpp/src/parquet/arrow/schema_internal.cc]: > Here, every caller except here is set to true. See: https://github.com/apache/arrow/blob/main/cpp/src/parquet/schema.cc#L443
> 
> By the way, I didn't find Int96 in the doc, would Int96 should having `is_adjusted_to_utc = true` ? (Rust impl set it false)

IMO, INT96 is a physical type and orthogonal to logical/converted type.

File: cpp/src/parquet/types.cc (hunk 2 above)
Comment (wgtmac) [reply cpp/src/parquet/types.cc:347]: It would be good to add the parity for timestamp type as well: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-timestamp-convertedtype, just in case people are looking for timestamp type only.

File: cpp/src/parquet/arrow/schema_internal.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/arrow/schema_internal.cc]: > INT96 is a physical type and orthogonal to logical/converted type.

Yeah, currently when we meet int96, we will bindly casting it to time-related, `!adjustedToUtc` value

File: cpp/src/parquet/types.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/parquet/types.cc:347]: done

**Review by wgtmac — APPROVED** [review]

File: cpp/src/parquet/arrow/schema_internal.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/arrow/schema_internal.cc]: @joellubi :

https://github.com/apache/arrow/blob/main/cpp/src/parquet/schema.cc#L443 The thrift schema would be casted here, and then goes to `LogicalType::FromConvertedType`. It would set `adjustedToUtc = true`: https://github.com/apache/arrow/blob/main/cpp/src/parquet/types.cc#L347-L358

So I think just `timestamp.is_adjusted_to_utc()` is ok

**Review by pitrou — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
The top post of the PR is not very explicit about it, so just mentioning it here as well (I posted an example on the issue): we need to be aware this changes the compatibility with legacy files written by Arrow. 

The fact that we ignored `adjustedToUtc` when constructing the Arrow TimestampType from a column that only used Parquet's legacy ConvertedType, was explicitly added as a way to ensure older files written by Arrow would still read correctly (https://github.com/apache/arrow/issues/22303).

Above it speaks about reading legacy files _not_ written by Arrow:

> So when it read legacy Parquet file not written by arrow, it would not have right type.

While this indeed "fixes" reading of non-Arrow legacy files (following the Parquet spec), it does change behaviour also for files written _by Arrow_. And for those files we knew that the original Arrow data used for writing could have timestamps with or without timezone, and at the time we decided that the safe option was to "refuse to guess", and thus default to not attaching the UTC timezone to the read data.

Nowadays it is maybe OK to make a different trade-off, and to prioritize the Parquet spec / non-Arrow legacy files above compatibility of reading Arrow legacy files, given that it is about files written by an Arrow version of several years old (Arrow < 0.15). But it should be a conscious decision (and labeled as a breaking change?).

**Comment by mapleFU** [comment]
> Nowadays it is maybe OK to make a different trade-off, and to prioritize the Parquet spec / non-Arrow legacy files above compatibility of reading Arrow legacy files, given that it is about files written by an Arrow version of several years old (Arrow < 0.15). But it should be a conscious decision (and labeled as a breaking change?).

I think should we add a config in `ArrowReaderProperties`, and shift to this impl in the coming versions?

**Comment by pitrou** [comment]
> Nowadays it is maybe OK to make a different trade-off, and to prioritize the Parquet spec / non-Arrow legacy files above compatibility of reading Arrow legacy files, given that it is about files written by an Arrow version of several years old (Arrow < 0.15).

If we really care about this, we should perhaps detect the offending Arrow version, since the Parquet file records the producer name and version. 

But we would also need an actual file to test with.
